File: 백준/Silver/11403. 경로 찾기/경로 찾기.py
from collections import deque

N = int(input())
adj_matrix = []
ans_matrix = [[0 for _ in range(N)]for __ in range(N)]

for _ in range(N):
    adj_matrix.append(list(map(int, input().split())))


def bfs(src, adj_matrix, ans_matrix):
    visit = [0 for _ in range(N)]
    visit[src] = 1
    queue = deque([src])
    while queue:
        cur = queue.popleft()
        for i in range(N):
            if adj_matrix[cur][i] and not visit[i]:
                visit[i] = 1
                ans_matrix[src][i] = 1
                queue.append(i)
            if adj_matrix[cur][i] and visit[i] and cur != src and i == src:
                ans_matrix[src][src] = 1


for i in range(N):
    bfs(i, adj_matrix, ans_matrix)

for i in range(N):
    for j in range(N):
        print(ans_matrix[i][j], end=" ")
    print()

File: 백준/Silver/11403. 경로 찾기/경로 찾기.java
import java.util.ArrayDeque;
import java.util.Collection;
import java.util.Queue;
import java.util.Scanner;

public class Main {
    static int[] visit;
    static int[] ans;
    static int[][] adj;
    static int N;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        adj =  new int[N][N];

        for (int i=0; i<N; i++){
            for (int j=0; j<N; j++){
                adj[i][j]= sc.nextInt();
            }
        }
        for (int i=0; i<N; i++){
            visit = new int[N];
            ans = new int[N];

            bfs(i);
            for (int j=0; j<N; j++){
                System.out.print(ans[j]+" ");
            }
            System.out.println();
        }
    }

    private static void bfs(int start) {
        visit[start] = 1;
        Queue<Integer> queue = new ArrayDeque<>();
        queue.add(start);

        while (!queue.isEmpty()){
            int cur = queue.poll();
            for (int i=0; i<N; i++){
                //길이 없다면 패스
                if(adj[cur][i]==0){
                    continue;
                }
                //이미 방문한적있고, 시작점이 아니라면 패스
                else if (visit[i]==1 && i!=start) {
                    continue;
                }
                //이미 방문한적있고, 시작점이라면
                else if (visit[i]==1 && i==start) {
                    ans[i]=1;
                }
                //방문한 적이 없다면
                else {
                    visit[i] = 1;
                    ans[i] = 1;
                    queue.add(i);
                }
            }
        }
    }
}